store/postgres/src/notification_listener.rs

```diff
@@ -255,11 +255,27 @@ impl JsonNotification {
 
             // Delete any notifications older than 5 minutes; this serves
             // as a regular cleanup so the table doesn't grow in size indefinitely
-            diesel::sql_query(
-                "DELETE FROM large_notifications
-                 WHERE created_at < current_timestamp - interval '300s'",
-            )
-            .execute(conn)?;
+            //
+            // We make sure that we do not wait for any locks when we do the
+            // cleanup to avoid stalling this transaction. There is not much
+            // danger in this cleanup sometimes leaving rows behind, as long
+            // as they get cleaned up eventually
+
+            // 'delete' locks the table in 'row exclusive' mode; see that we
+            // can get that lock without waiting
+            let lock =
+                diesel::sql_query("lock table large_notifications in row exclusive mode nowait")
+                    .execute(conn);
```

Comment:
Does this keep the lock after the execute? Maybe the lock and the delete should be in the same transaction?

Reply:
This code is run when we already are in a transaction, and Postgres doesn't support autonomous transactions. Yes, the lock is kept until the end of the overall txn (that happens for all locks, they only get released when the txn commits or rolls back)

All that lock statement is doing is explicitly acquiring the lock that the `delete` would acquire implicitly, with the only difference being that `delete` would wait for this lock, and the above code reports an error if it can't get the lock immediately, in which case we don't even try to delete.